### DISCUSSION TITLE
Declare last callbacks are executed multiple times

### DISCUSSION THREAD

**Discussion by DavideCanton** [sqlalchemy/sqlalchemy#9830] [Usage Questions] [answered]
The following code:

```py
from sqlalchemy.orm import DeclarativeBase, mapped_column, Mapped


class Base1(DeclarativeBase):
    pass


class Model1(Base1):
    id: Mapped[int] = mapped_column(primary_key=True)
    __tablename__ = "t1"

    @classmethod
    def __declare_last__(cls):
        print("foo")


class Base2(DeclarativeBase):
    pass


class Model2(Base2):
    id: Mapped[int] = mapped_column(primary_key=True)
    __tablename__ = "t2"

    @classmethod
    def __declare_last__(cls):
        print("bar")


Base1.registry.configure()
Base2.registry.configure()
```

runs the `__declare_last__` callbacks multiple times, as the output of this script is
```shell
$ python foo.py
foo
bar
foo
bar
```

Maybe I'm missing something, but since the two base classes are not related to each other, I was expecting each registry to be configured once.

I've tried with `sqlalchemy==2.0.4` and `sqlalchemy==2.0.15`, with the same result.

**Comment by zzzeek** [comment, marked as answer]
that's a good idea which we should make possible for 2.1 (cc @CaselIT ) but that is not how things are architected right now.    The idea of there being multiple registries that are independently configurable was a new idea introduced in SQLAlchemy 1.4, and when that was done, we have not yet added new event hooks that are local to specific registries.    Right now ``__declare_last__`` is against the `Mapper` wide [after_configured](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.MapperEvents.after_configured) event and there are no event hooks as yet that are local to registries, like `after_registry_configured`, which we absolutely should add.   The ``__declare_last__`` and ``__declare_first__`` "convenience" hooks should allow some pathway to be registry local, although I dont know if we should make it do that by default, or at least have some way to switch between the two behaviors.   Easiest way would be just to make new hooks like ``__declare_registry_first__``, ``__declare_registry_last__``.

anyway, that's all things we haven't thought about yet until now so for the moment you need to set up your own state to work around things.

**Reply by DavideCanton** [reply]
Ok, I've noticed this behaviour since I had a relationship declared inside a `__declare_last__` function which references another model in a circular fashion, and a warning was raised because that relationship was overwritten (it was a particular test setup anyway, so nothing particularly problematic).

**Reply by DavideCanton** [reply]
My two cents: probably having two sets of hooks is easier to understand for the user

**Comment by DavideCanton** [comment]
Regarding the problem that I mentioned in my comment, (I know this is a totally unrelated question, so I can move it to a different discussion), I've noticed, thanks to the `__declare_last__` multiple invocations, that column properties set multiple times on a mapped class are still referenced in the mapped column collection.

For example
```py
class Model(Base):
    id: Mapped[int] = mapped_column(primary_key=True)

    __tablename__ = "table"


Model.foo = column_property(Model1.id > 0)
Model.foo = column_property(Model1.id < 0)

assert len(inspect(Model).columns) == 3  # there are two copies of foo in the collection
assert len(inspect(Model).iterate_properties) == 2  # properties are stored in a dict, so there are no duplicates here
assert len(select(Model).columns) == 2  # the select correctly retrieves only one column property
```

I don't know if this is intended behavior or not, to me it feels a bit counterintuitive.

**Reply by zzzeek** [reply]
it's undefined behavior.  IMO it should raise an error, since removal of individual attributes is not supported.

**Reply by DavideCanton** [reply]
I think it now raises a warning in some cases

**Reply by zzzeek** [reply]
it does not.  there's a related warning that it's trying to combine the columns together in some cases but otherwise right now there is no check for this at all - #9841

**Reply by DavideCanton** [reply]
I'm referring to [this](https://github.com/sqlalchemy/sqlalchemy/blob/2dced78f33010d4a6a6276f4e8c5665bd08d01c0/lib/sqlalchemy/orm/mapper.py#L2180), that's how I got that the callbacks were invoked many times

**Reply by zzzeek** [reply]
it looks like that warning would only be emitted for a relationship right now.

this patch might be the fix for now

```diff
diff --git a/lib/sqlalchemy/orm/mapper.py b/lib/sqlalchemy/orm/mapper.py
index 0533c82a69..825a8c31ef 100644
--- a/lib/sqlalchemy/orm/mapper.py
+++ b/lib/sqlalchemy/orm/mapper.py
@@ -2403,7 +2403,7 @@ class Mapper(
         the given MapperProperty is configured immediately.
 
         """
-        prop = self._configure_property(key, prop, init=self.configured)
+        prop = self._configure_property(key, prop, init=self.configured, warn_for_existing=True)
         assert isinstance(prop, MapperProperty)
         self._init_properties[key] = prop
 

```

**Reply by DavideCanton** [reply]
Oh ok, it was on a relationship, not on a column property.

Also watch out that it seems to emit two warnings on relationships with your patch:

```py
from sqlalchemy import ForeignKey
from sqlalchemy.orm import *

from db import Base


class Model(Base):
    id: Mapped[int] = mapped_column(primary_key=True)

    __tablename__ = "table"


class Model2(Base):
    id: Mapped[int] = mapped_column(ForeignKey("table.id"), primary_key=True)
    __tablename__ = "table2"


Model.bar = relationship("Model2", primaryjoin=Model.id == Model2.id)
Model.bar = relationship("Model2", primaryjoin=Model.id == Model2.id)

Base.registry.configure()
```
outputs:
```
foo.py:21: SAWarning: Property Model.bar on Mapper[Model(table)] being replaced with new property Model.bar; the old property will be discarded
  Model.bar = relationship("Model2", primaryjoin=Model.id == Model2.id)
foo.py:21: SAWarning: User-placed attribute <sqlalchemy.orm.attributes.InstrumentedAttribute object at 0x00000220C0581B20> on Mapper[Model(table)] being replaced with new property "Model.bar"; the old attribute will be discarded
  Model.bar = relationship("Model2", primaryjoin=Model.id == Model2.id)
```

**Reply by zzzeek** [reply]
thanks.  yes, the patch got more complicated after running some tests, you can see current status at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4633
